test/unit_test/api/apps/services/test_provider_api_service_verify_discovery.py

```
@@ -0,0 +1,238 @@
+    assert redact("https://host/v1#tok=SECRET") == "https://host/v1"
+    assert redact("host/v1?api-key=SECRET") == "<unparsable url>"
+    assert redact("https://host:notaport/v1") == "<unparsable url>"
+    assert redact("") == ""
+    assert redact(None) == ""
+
+
+def _load_helper(name: str):
+    repo_root = Path(__file__).resolve().parents[5]
+    source = (repo_root / "api" / "apps" / "services" / "provider_api_service.py").read_text()
+    namespace: dict = {}
+    exec(compile(_extract_function(source, name), "provider_api_service.py", "exec"), namespace)
```

Ruff (0.16.1)

[error] Use of `exec` detected

(S102)